Sources/Basics/HTTPClient.swift

```diff
@@ -0,0 +1,334 @@
+}
+
+public struct HTTPClientRequest {
+    let method: Method
+    let url: URL
+    var headers: HTTPClientHeaders
+    var body: Data?
+    var options: Options
+
+    public init(method: Method = .get,
+                url: URL,
+                headers: HTTPClientHeaders = .init(),
+                body: Data? = nil,
+                options: Options = .init()) {
+        self.method = method
+        self.url = url
+        self.headers = headers
+        self.body = body
+        self.options = options
+    }
+
+    public enum Method {
+        case head
+        case get
+        case post
+        case put
+        case delete
+    }
+
+    public struct Options {
+        var addUserAgent: Bool
+        var validResponseCodes: [Int]?
+        var timeout: DispatchTimeInterval?
+        var retryStrategy: HTTPClientRetryStrategy?
+        var circuitBreakerStrategy: HTTPClientCircuitBreakerStrategy?
+        var callbackQueue: DispatchQueue?
+
+        public init() {
+            self.addUserAgent = true
+            self.validResponseCodes = .none
+            self.timeout = .none
+            self.retryStrategy = .none
+            self.circuitBreakerStrategy = .none
+            self.callbackQueue = .none
+        }
+    }
+}
+
+public struct HTTPClientResponse {
+    let statusCode: Int
+    let statusText: String?
+    let headers: HTTPClientHeaders
+    let body: Data?
+
+    public init(statusCode: Int,
+                statusText: String? = nil,
+                headers: HTTPClientHeaders = .init(),
+                body: Data? = nil) {
+        self.statusCode = statusCode
+        self.statusText = statusText
+        self.headers = headers
+        self.body = body
+    }
+}
+
+public struct HTTPClientHeaders: Sequence, Equatable {
+    private var items: [Item]
+    private var headers: [String: [String]]
+
+    public init(_ items: [Item] = []) {
+        self.items = items
+        self.headers = items.reduce([String: [String]]()) { partial, item in
+            var map = partial
+            // Avoid copy-on-write: remove entry from dictionary before mutating
+            var values = map.removeValue(forKey: item.name.lowercased()) ?? []
+            values.append(item.value)
+            map[item.name.lowercased()] = values
+            return map
+        }
+    }
+
+    public func contains(_ name: String) -> Bool {
+        self.headers[name.lowercased()] != nil
+    }
+
+    public var count: Int {
+        self.headers.count
+    }
+
+    public mutating func add(name: String, value: String) {
+        self.add(Item(name: name, value: value))
+    }
+
+    public mutating func add(_ item: Item) {
+        // Avoid copy-on-write: remove entry from dictionary before mutating
+        var values = self.headers.removeValue(forKey: item.name.lowercased()) ?? []
+        values.append(item.value)
+        self.headers[item.name.lowercased()] = values
+        self.items.append(item)
```

Comment:
If I read this correctly, if the same item is added multiple times, this would result in duplicates in the array — is that intentional?

Reply:
http headers are really [string: [string]] i.e. having multiple entries for the same header name is valid. that said you are right that this allows pure duplicates.